spec.html

```diff
@@ -13695,8 +13695,9 @@ <h1>Runtime Semantics: Evaluation</h1>
         <emu-alg>
           1. Let _newTarget_ be GetNewTarget().
           1. Assert: Type(_newTarget_) is Object.
-          1. Let _func_ be ? GetSuperConstructor().
+          1. Let _func_ be ! GetSuperConstructor().
           1. Let _argList_ be ? ArgumentListEvaluation of |Arguments|.
+          1. If IsConstructor(_func_) is *false*, throw a *TypeError* exception.
```

Comment:
this might need to verify if `_func_` ir a normal or an abrupt completion. GetSuperConstructor will do the `IsConstructor` check.

Reply:
Basically `ReturnIfAbrupt(_func_).`

Reply:
Doesn't this PR remove GetSuperConstructor from ever returning an abrupt completion?

Reply:
GetSuperConstructor can still return an abrupt completion, but it's held in these steps.

The abstraction will also check if _func_ is a constructor and throw TypeError otherwise, so `_func_` can be it with this change.

Maybe what you want is even more simple, like:


```diff
-          1. Let _func_ be ? GetSuperConstructor().
          1. Let _argList_ be ? ArgumentListEvaluation of |Arguments|.
+          1. Let _func_ be ? GetSuperConstructor().
```

Moving the GetSuperConstructor check to remain after the ArgumentListEvaluation should resolve the problem.
_func_

Reply:
Just a matter of an observable `Let superConstructor be ! activeFunction.[[GetPrototypeOf]]().` happening after ArgumentListEvaluation. (you basically detect that with Proxies)

Reply:
After some conversation on irc, I drafted my suggestion with code examples: https://gist.github.com/leobalter/17ba75d682094fb028f6a6e05f82f501

As pointed out by @devsnek, it's easy to try out this change in engine262: https://github.com/engine262/engine262/blob/master/src/runtime-semantics/SuperCall.mjs#L33

Reply:
This seems ok. I was concerned with revocation of super constructor proxies inside arguments but don't see any issues with them upon closer inspection.

Reply:
the revocation in the super arguments would be somehow similar to:

```js
var X = Proxy.revocable(function() {}, {});
new (X.proxy)(X.revoke());
```

Reply:
I think I prefer the current PR (*), because it keeps the left-to-right evaluation order (**), i.e. `super` is evaluated before the arguments, similar to how `f` in `f((f = g))` is evaluated before the assignment in the arguments expression. 

(*) Move the `IsConstructor` out of `GetSuperConstructor`, making `GetSuperConstructor` an infallible operation. And then performing `IsConstructor` right before `Construct`, similar to how `EvaluateNew` works. 
(**) Normally left-to-right evaluation is preferred, cf. the on-going discussion about [property accessor](https://tc39.es/ecma262/#sec-property-accessors) evaluation and implementation reality. 

---

In addition to keeping the left-to-right semantics, the current PR is also slightly more efficient to implement (in SpiderMonkey at least), because it avoids additional stack juggling to move `func` in the correct stack position. But that's just an added benefit and shouldn't drive the specification process. :smile:

Reply:
This seems to come down to whether a non-constructor `super(…)` is considered analogous to a non-function TypeError (e.g., `null(console.log("works"))` or an uninitialized-binding ReferenceError (e.g., `tdz(console.log("fails")); let tdz = x=>x`. But I will note that like the former case and unlike the latter, this situation is fixable at runtime:
```
$ eshost -si proof.js 
## Source
class A extends null {
  constructor( base ) {
    Object.setPrototypeOf(A, base);
    super();
  }
}
print(new A(class {}) && "ok");


#### Chakra, JavaScriptCore, SpiderMonkey, V8, XS
ok
```

Reply:
@leobalter I am pretty sure that `GetSuperConstructor` cannot throw. In your code snippet

```js
var x = 0;
var getproto = false;
class C extends Object { // NB I changed this from `null` because `extends null` is confusing
  constructor() {
    super(x = 1);
  }
}

var P = new Proxy(C, {
  getPrototypeOf() {
    getproto = true;
    return null;
  }
});

try {
  new P;
} catch {}
```

the `getPrototypeOf` trap is never actually invoked. The only place that it would possibly be invoked is in the `[[GetPrototypeOf]]` in `GetSuperConstructor`, but at that point the active function object is `C`, not `P`. (The `[[Construct]]` internal method for proxies like this one just calls the target's `[[Construct]]` directly; it doesn't try to invoke it in a way which would override the active function object somehow.)

So whether the `GetSuperConstructor` happens before or after argument list evaluation is not observable.

Editorially speaking, I prefer it to happen before argument list evaluating (as in this PR), because I think it makes sense to have the order of operations for `super(args)` be "have all of the effects of evaluating `super`, then all of the effects of evaluating the arguments, then check for constructibility, then invoke", just like the order of operations for `new foo(args)` is "have all of the effects of evaluating `foo`, then all of the effects of evaluating the arguments, then check for constructibility, then invoke".

Reply:
That suggests, then, that this PR is ready to go and needs no further changes, here or in test262.

Can some folks confirm? I'd love to land this soon.

Reply:
@ljharb

> this PR is ready to go and needs no further changes

that's my position. Would be good to check with @leobalter too.

Reply:
is it valid to extend ecmascript to have a kind of function where [[GetPrototypeOf]] on it could return abruptly?

Reply:
@devsnek Nothing forbids such an extension, to my knowledge, but `GetSuperConstructor` is only invoked by functions created by `class`, which are not such functions. (Hence the `Assert: activeFunction is an ECMAScript function object`.)